File: IT/2/Multithreading/Esercizi/C++/esercizio28.cpp
/*
file esercizio28.cpp
compilare con: g++ -std=c++11 esercizio28.cpp -lpthread -I.
oppure (c++20):
g++ -std=c++20 esercizio28.cpp -lpthread
*/

#include <random>
#include <thread>
#include <mutex>
#if __cplusplus > 201703L
#include <semaphore>
using std::counting_semaphore;
#else
#include "semaphore.h"
#endif
using semaphore = counting_semaphore<1>;

using std::random_device;
using std::default_random_engine;
using std::uniform_int_distribution;
using std::thread;
using std::mutex;
using std::chrono::duration;
using std::chrono::seconds;
using std::ref;

void barber(semaphore& barberReady, semaphore& accessWRSeats, semaphore& customerReady, int& numberOfFreeWRSeats, int* customers, int i) {
	thread_local default_random_engine gen(random_device{}());

	while (true) {
		customerReady.acquire();
		accessWRSeats.acquire();
		int pc = customers[numberOfFreeWRSeats];
		numberOfFreeWRSeats += 1;
		barberReady.release();
		accessWRSeats.release();

		printf("Il barbiere thread %d sta tagliando i capelli al cliente thread %d.\n", i, pc);
		uniform_int_distribution<int> dist(10, 25);
		int pause = dist(gen);
		seconds duration(pause);
		std::this_thread::sleep_for(duration);

		printf("thread %d ha impiegato %d minuti per tagliare i capelli a thread %d. Avanti un altro!\n", i, pause, pc);
	}
}

void customer(semaphore& barberReady, semaphore& accessWRSeats, semaphore& customerReady, int& numberOfFreeWRSeats, int* customers, int i) {
	thread_local default_random_engine gen(random_device{}());

	printf("Il cliente thread %d entra nel negozio.\n", i);
	accessWRSeats.acquire();

	if (numberOfFreeWRSeats > 0) {
		numberOfFreeWRSeats -= 1;
		customers[numberOfFreeWRSeats] = i;
		printf("thread %d si siede e aspetta di farsi tagliare i capelli...\n", i);
		customerReady.release();
		accessWRSeats.release();
		barberReady.acquire();
	}
	else {
		printf("Il cliente thread %d non ha trovato posto per sedersi ed esce dal negozio...\n", i);
		customerReady.release();
	}

	uniform_int_distribution<int> dist(1, 25);
	seconds duration(dist(gen));
	std::this_thread::sleep_for(duration);
}

int main()
{
	semaphore barberReady(0);
	semaphore accessWRSeats(1);
	semaphore customerReady(0);
	int numberOfFreeWRSeats = 5;

	int customers[5] = {0, 0, 0, 0, 0};
	int i = 1;

	thread tbarber(barber, ref(barberReady), ref(accessWRSeats), ref(customerReady), ref(numberOfFreeWRSeats), customers, i);

	thread processes[25];

	while (true) {
		i++;
		processes[i] = thread(customer, ref(barberReady), ref(accessWRSeats), ref(customerReady), ref(numberOfFreeWRSeats), customers, i);
		thread_local default_random_engine gen(random_device{}());
		uniform_int_distribution<int> dist(5, 30);
		seconds duration(dist(gen));
		std::this_thread::sleep_for(duration);
	}

	tbarber.join();

	for (auto& c : processes) {
		c.join();
	}
}

File: IT/2/Multithreading/Esercizi/Python/esercizio28.py
# file esercizio28.py

from __future__ import print_function
from multiprocessing import Process, current_process, Semaphore, Value, Manager
from random import randint
from time import sleep

def barber(barberReady, accessWRSeats, customerReady, numberOfFreeWRSeats, customers):
	p = current_process().name
	while True:
		customerReady.acquire()
		accessWRSeats.acquire()

		pc = customers[numberOfFreeWRSeats.value]
		numberOfFreeWRSeats.value += 1
		barberReady.release()
		accessWRSeats.release()

		print("Il barbiere {} sta tagliando i capelli al cliente {}.".format(p, pc))
		pause = randint(10, 25)
		sleep(pause)
		print("{} ha impiegato {} minuti per tagliare i capelli a {}. Avanti un altro!".format(p, pause, pc))

def customer(barberReady, accessWRSeats, customerReady, numberOfFreeWRSeats, customers):
	p = current_process().name
	print("Il cliente {} entra nel negozio.".format(p))

	accessWRSeats.acquire()
	if numberOfFreeWRSeats.value > 0:
		numberOfFreeWRSeats.value -= 1
		customers[numberOfFreeWRSeats.value] = p
		print("{} si siede e aspetta di farsi tagliare i capelli...".format(p))
		customerReady.release()
		accessWRSeats.release()
		barberReady.acquire()
	else:
		print("Il cliente {} non ha trovato posto per sedersi ed esce dal negozio...".format(p))
		accessWRSeats.release()

if __name__ == "__main__":
	barberReady = Semaphore(0)
	accessWRSeats = Semaphore(1)
	customerReady = Semaphore(0)
	numberOfFreeWRSeats = Value('i', 5)

	# lista condivisa tra i processi tramite un manager
	customers = Manager().list([""]*5)

	pbarber = Process(target=barber, args=(barberReady, accessWRSeats, customerReady, numberOfFreeWRSeats, customers,))

	pbarber.start()

	procs = []

	while True:
		pcustomer = Process(target=customer, args=(barberReady, accessWRSeats, customerReady, numberOfFreeWRSeats, customers,))
		pcustomer.start()
		procs.append(pcustomer)
		pause = randint(5, 30)
		sleep(pause)

	pbarber.join()

	for c in procs:
		c.join()

